### ISSUE TITLE
adding a task takes long time

### ISSUE THREAD

**Issue by meguiraun** [mxcube/mxcube3#738]
Adding a single task takes no time, but the time is increased the bigger the queue is.

It seems that on `Queue.queue_add_item` we return the whole queue, and then on the client we re-apply on `actions.queue.setQueue`. And we recreate the whole queue on each call.

This has something to do with the interleaving but I cannot figure out why. Any idea?

**Comment by marcus-oscarsson** [comment]
Well its not directly related to the interleaved but it makes the implementation simpler.

We return the entire queue to the client when we add a task (any task). so add_task(task) => queue. This is not a problem in it self but the code that checks for file collisions is something like O(n2), so the time for the operation of adding a task will always (with the current implementation) grow with the number of tasks. Disk access is also relatively slow which does not help. Our intention have been to fix this and I have started to prepare for a faster way of doing these file checks.

Serializing the queue is of-course also dependent on the size of the queue, but basically just converts a number of dicts to json, so that should be very fast. I say should because it might be that we are reading some hardware or LIMS attribute which would make that operation considerably much slower.

So two things to consider for the performance, 

    1. Faster file collision detection (we have already a start implemented in MXCuBE3 so let me know if you intend to continue with that so we can discuss it before)

    2. Make sure that you/we are not reading value or LIMS in the queue serialization code

Finally, why is it like this ? So as you say it has something to do with the interleaved. When a user selects n tasks to interleave we like to concatenate those tasks into one task. Whats happening is basically:

1. interleaved = create_interleaved_from(n1, n2, ...n), 
2. remove(n1, n2, .., n)
3. insert(interleaved, positon_where_n1_was)

This is as you know done on the server and the client does not really now what an interleaved is, so the easiest thing is to simply return the entire queue after each operation that adds or removes items. Not the fastest, most elegant or even interesting one, just simple :). Although it should (should) have little impact on the time it takes to add tasks.

**Comment by marcus-oscarsson** [comment]
I have disabled mxcube.queue.check_for_path_collisions(pt) here for the time being since we now have automatic run number generation that should handle the problem. It should be noted that the get_run_number function also can be optimized.

Re enabling  mxcube.queue.check_for_path_collisions(pt) I notice that it takes considerably much longer time to add tasks, so maybe try without that.

**Comment by marcus-oscarsson** [comment]
So I became curious and looked a bit at the code, the call that takes time for us here is: 

  ```limsres = mxcube.rest_lims.get_dc(lims_id)```

I added a parameter a while a go to disable this by passing include_lims_data = False. So try to
set that to False and see if it speeds things up. I've made the necessary changes here to make sure that include_lims_data is passed correctly when serializing the queue (Not committed yet). We basically only need to pass True when the user refreshes the page.

Let me know if it helps you or if you have some other ideas that might help. Otherwise for  use here:

1. Disabling the outdated file collision detection
2. Passing inlcude_lims_data = False | True at the right moments

Makes a huge difference

**Comment by meguiraun** [comment]
okey thanks a lot for the clarification.

I saw the check for path collisions, but I did not see any noticeable delay there. Here it is the queue_to_dict_rec code that takes most of the time. 

Anyway, I will disable the checks and test again.

In any case, I do not like much the fact that we are sending back the whole queue when adding a single task. Not suggesting to change it now but we need to consider.
